### PR TITLE
[SPARK-43261][PYTHON] Migrate `TypeError` from Spark SQL types into error class.

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to migrate `TypeError` from Spark SQL types into error class.

### Why are the changes needed?

To improve PySpark error

### Does this PR introduce _any_ user-facing change?

No API change, only error improvement.

### How was this patch tested?

The existing CI should pass

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -2183,7 +2216,13 @@ def asDict(self, recursive: bool = False) -> Dict[str, Any]:
         True
         """
         if not hasattr(self, "__fields__"):
-            raise TypeError("Cannot convert a Row class into dict")
+            raise PySparkTypeError(
+                error_class="CANNOT_CONVERT",
+                message_parameters={
+                    "from_type": "Row",
```
Comment (allisonwang-db) [inline python/pyspark/sql/types.py:2222]: Here should be `a Row class`?

File: python/pyspark/errors/error_classes.py [hunk 2]
```diff
@@ -74,6 +84,21 @@
       "Can not infer Array Type from an list with None as the first element."
     ]
   },
+  "CANNOT_INFER_SCHEMA_FOR_TYPE": {
+    "message": [
+      "Can not infer schema for type: `<data_type>`."
```
[the remark this answers is not in the document]
Comment (itholic) [reply python/pyspark/errors/error_classes.py:89]: Yes, and this behavior just simply mimics JVM-side error messages.

**Comment by itholic** [comment]
CI passed. @zhengruifeng Could you take an another look when you find some time?

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to master

**Pull request closed without merging**